usr.bin/login/common.c

```diff
@@ -0,0 +1,418 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <sys/cdefs.h>
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/socket.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <syslog.h>
+#include <fcntl.h>
+#ifdef NOT_IMPLEMENTED
+#include <ttyent.h>
+#endif
+#include <setjmp.h>
+#include <time.h>
+#include <pwd.h>
+#include <err.h>
+#include <vis.h>
+#include <util.h>
+#include <signal.h>
+
+#include "pathnames.h"
+#include "common.h"
+
+#if defined(KERBEROS5)
+#define	NBUFSIZ		(MAXLOGNAME + 1 + 5)	/* .root suffix */
+#else
+#define	NBUFSIZ		(MAXLOGNAME + 1)
+#endif
+
+#ifdef SUPPORT_UTMP
+#include <utmp.h>
+static void	 doutmp(void);
+static void	 dolastlog(int);
+#endif
+#ifdef SUPPORT_UTMPX
+#include <utmpx.h>
+static void	 doutmpx(void);
+static void	 dolastlogx(int);
+#endif
+
+/*
+ * This bounds the time given to login.  Not a define so it can
+ * be patched on machines where it's too small.
+ */
+u_int	timeout = 300;
+
+#ifdef NOT_IMPLEMENTED
+void	 decode_ss(const char *);
+#endif
+struct	passwd *pwd;
+int	failures, have_ss;
+char	term[64], *envinit[1], *hostname, *tty, *nested;
+const char *username;
+struct timeval now;
+#ifdef NOT_IMPLEMENTED
+struct sockaddr_storage ss;
+#endif
+
+char *
+trimloginname(char *u)
+{
+	if (strlen(u) > MAXLOGNAME)
+		u[MAXLOGNAME] = '\0';
+	return u;
+}
+
+char *
+getloginname(void)
+{
+	int ch;
+	char *p;
+	static char nbuf[NBUFSIZ];
+
+	for (;;) {
+		(void)printf("login: ");
+    fflush(stdout);
```

Comment:
@j-piecuch Could you help @franciscozdo investigating why `fflush` is needed here, but NetBSD does without it? Maybe we're missing something in `tty` driver?

Reply:
I fixed it in #962. `fstat` during `__smakebuf()` was failing